tests/unit/s2n_async_offload_cb_test.c

```diff
@@ -0,0 +1,242 @@
+#include "tls/s2n_security_policies.h"
+#include "utils/s2n_safety.h"
+
+struct s2n_async_offload_cb_test {
+    unsigned invoke_perform : 1;
+    int result;
+
+    int invoked_count;
+    struct s2n_async_op *op;
+};
+
+int s2n_async_offload_test_callback(struct s2n_connection *conn, struct s2n_async_op *op, void *ctx)
+{
+    EXPECT_NOT_NULL(op);
+
+    struct s2n_async_offload_cb_test *data = (struct s2n_async_offload_cb_test *) ctx;
+    data->invoked_count += 1;
+    data->op = op;
+
+    if (data->invoke_perform) {
+        EXPECT_SUCCESS(s2n_async_op_perform(op));
+    }
+
+    return data->result;
+}
+
+static int s2n_test_handshake_async(struct s2n_connection *server_conn, struct s2n_connection *client_conn,
+        struct s2n_async_offload_cb_test *data)
+{
+    while (true) {
+        int ret_val = s2n_negotiate_test_server_and_client(server_conn, client_conn);
+
+        if (s2n_errno == S2N_ERR_ASYNC_BLOCKED) {
+            /* Handshake remains blocked as long as op_perform() is not invoked. */
+            for (int i = 0; i < 3; i++) {
+                EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server_conn, client_conn),
+                        S2N_ERR_ASYNC_BLOCKED);
+            }
+
+            EXPECT_SUCCESS(s2n_async_op_perform(data->op));
+            /* Each operation can only be performed once. */
+            EXPECT_FAILURE_WITH_ERRNO(s2n_async_op_perform(data->op), S2N_ERR_INVALID_STATE);
+        } else {
+            return ret_val;
+        }
+    }
+    return 0;
+}
+
+int main(int argc, char *argv[])
+{
+    BEGIN_TEST();
+
+    /* Safety Check */
+    {
+        struct s2n_async_offload_cb_test test_data = { 0 };
+        EXPECT_FAILURE_WITH_ERRNO(s2n_config_set_async_offload_callback(NULL, s2n_async_offload_test_callback, S2N_ASYNC_ALLOW_ALL, &test_data),
+                S2N_ERR_NULL);
+
+        DEFER_CLEANUP(struct s2n_config *test_config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_NOT_NULL(test_config);
+        EXPECT_EQUAL(test_config->async_offload_allow_list, S2N_ASYNC_OP_NONE);
+        EXPECT_FAILURE_WITH_ERRNO(s2n_config_set_async_offload_callback(test_config, NULL, S2N_ASYNC_ALLOW_ALL, &test_data),
+                S2N_ERR_NULL);
+
+        EXPECT_SUCCESS(s2n_config_set_async_offload_callback(test_config, s2n_async_offload_test_callback,
+                S2N_ASYNC_PKEY_VERIFY, &test_data));
+        EXPECT_TRUE(s2n_async_is_op_in_allow_list(test_config, S2N_ASYNC_PKEY_VERIFY));
+
+        EXPECT_FAILURE_WITH_ERRNO(s2n_async_op_perform(NULL), S2N_ERR_NULL);
+        struct s2n_async_op test_op = { 0 };
+        EXPECT_FAILURE_WITH_ERRNO(s2n_async_op_perform(&test_op), S2N_ERR_INVALID_STATE);
+    }
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
```

Comment:
Nit: I'd put this at the very top of main, before the safety test. That's where future devs will look for it if they also need a cert chain.